edgeless_node/src/resources/ollama.rs

```diff
@@ -0,0 +1,344 @@
+        }
+    }
+}
+
+pub struct OllamaResourceProviderInner {
+    resource_provider_id: edgeless_api::function_instance::InstanceId,
+    dataplane_provider: edgeless_dataplane::handle::DataplaneProvider,
+    instances: std::collections::HashMap<edgeless_api::function_instance::ComponentId, OllamaResource>,
+    sender: futures::channel::mpsc::UnboundedSender<OllamaCommand>,
+    _handle: tokio::task::JoinHandle<()>,
+}
+
+pub struct OllamaResource {
+    join_handle: tokio::task::JoinHandle<()>,
+}
+
+impl Drop for OllamaResource {
+    fn drop(&mut self) {
+        self.join_handle.abort();
+    }
+}
+
+impl OllamaResource {
+    /// Create a new Ollama resource.
+    ///
+    /// - `resource_provider`: the resource provider.
+    /// - `dataplane_handle`: gives access to the EDGELESS dataplane.
+    /// - `model_name`: name of the AI model to use.
+    /// - `instance_id`: identifier of this resource instance.
+    /// - `sender`: channel to send commands to the resource task.
+    async fn new(
+        dataplane_handle: edgeless_dataplane::handle::DataplaneHandle,
+        model_name: String,
+        instance_id: edgeless_api::function_instance::InstanceId,
+        sender: futures::channel::mpsc::UnboundedSender<OllamaCommand>,
+    ) -> anyhow::Result<Self> {
+        let mut dataplane_handle = dataplane_handle;
+        let mut sender = sender;
+
+        let history_id = instance_id.function_id.to_string();
+        let handle = tokio::spawn(async move {
+            loop {
+                let edgeless_dataplane::core::DataplaneEvent {
+                    source_id,
+                    channel_id,
+                    message,
+                } = dataplane_handle.receive_next().await;
+
+                let mut need_reply = false;
+                let message_data = match message {
+                    edgeless_dataplane::core::Message::Call(data) => {
+                        need_reply = true;
+                        data
+                    }
+                    edgeless_dataplane::core::Message::Cast(data) => data,
+                    _ => {
+                        continue;
+                    }
+                };
+
+                let (reply_sender, reply_receiver) =
+                    tokio::sync::oneshot::channel::<anyhow::Result<(edgeless_api::function_instance::InstanceId, String)>>();
+                let _ = sender
+                    .send(OllamaCommand::Chat(
+                        model_name.clone(),
+                        history_id.clone(),
+                        message_data,
+                        instance_id.function_id.clone(),
+                        reply_sender,
+                    ))
+                    .await;
+
+                match reply_receiver.await {
+                    Ok(response) => match response {
+                        Ok((target, response)) => {
+                            let _ = dataplane_handle.send(target, response).await;
```

Comment:
Is there any reason to allow both cast and call here? Imho, this is rather confusing.

Reply:
Oh no, wrong copy/paste from the code of another resource. Will change that to only cast.